File: python/union_find.py
arr = [1,2,3,4,0,5]
par = [i for i in range(len(arr))]
rank = [1]* len(arr)

def find(node):
    node = par[node]
    while node!=par[node]:
        # path compression
        par[node] = par[par[node]]
        node = par[node]
    return node

def union(a,b):
    par_a,par_b = find(a),find(b)
    if par_a==par_b:
        return # no need to join
    # find the smaller group and merge it into larger group
    if rank[par_a] < rank[par_b]:
        rank[par_b] += rank[par_a]
        par[par_a] = par_b
    else:
        rank[par_a] += rank[par_b]
        par[par_b] = par_a
    return

File: Java/union-find.java
class DSU{
        int[] parent;
        int[] rank;

        DSU(int size){
            parent = new int[size];
            rank = new int[size];
            for(int i=0;i<size;i++)
                parent[i]=i;
            Arrays.fill(rank,1);
        }

        void union(int a, int b){
            int parentA = find(a);
            int parentB = find(b);
            if(parentA==parentB) return;

            if(rank[parentA]<rank[parentB]){
                parent[parentA]=parentB;
            }
            else if(rank[parentB]<rank[parentA]){
                parent[parentB]=parentA;
            }
            else{
                parent[parentA]=parentB;
                rank[parentB]++;
            }
        }

        int find(int a){
            if(a == parent[a]) return a;
            return parent[a]=find(parent[a]);
        }

    }
